### DISCUSSION TITLE
Unrealistic base drift

### DISCUSSION THREAD

**Discussion by alex-mitrevski** [hsr-project/general#16] [Q&A] [answered]
Hi,

While testing in the simulation, we noticed that the HSR drifts way too much when a velocity command is sent to the base. This is illustrated in the following video, where we send a small forward velocity of 0.02m/s; after a short while of moving forward, the robot starts rotating instead of actually moving forward:

https://user-images.githubusercontent.com/2042490/122960279-8d50ee80-d383-11eb-8c4a-5de8dc091c5f.mp4

I understand that the purpose of adding noise is to increase the realism of the simulation; however, the drift doesn't correspond at all to how the real HSR behaves (in particular, the drift on the real robot is more visible at higher velocities, but negligible for low velocities such as the one in the above example). If we can't rely at all on the accuracy of the robot motions, it's really difficult to do any tasks in the simulation properly.

Does anyone have any suggestions on how to counteract this issue?

Thanks.

**Comment by yosuke** [comment, marked as answer]
This may happen frequently if you go back and forth in a small area.
This is due to the special omnidrive mechanism used in the HSR base:
https://github.com/devrt/offset-diff-drive-controller

Official HSR uses laser odometry to reduce this issue but the current OSS implementation does not have this...

One possible workaround is to take a strategy to avoid small back and forth moves.
You can also use /map tf-frame to monitor your current position and attitudes.

**Reply by alex-mitrevski** [reply]
Thanks for the insights! We are indeed using the `/map` frame now and this has helped quite a bit (although I am careful when doing this since this assumes that the localisation is always accurate).
